Activation flow

beyang opened the pull request:
Addresses #975

beyang commented:
v1: ![image](https://user-images.githubusercontent.com/1646931/52616606-efad9200-2e4d-11e9-9c61-8bbb1e496914.png)

lguychard commented on web/src/site-admin/SiteAdminOnboarding.tsx:
```diff
@@ -0,0 +1,217 @@
+
+    public render(): JSX.Element {
+        const addCodeHosts = () => {
+            window.location.href = '/site-admin/external-services'
```
I know this is WIP, but you'll probably want to inject `history` into your component's props and to use `this.props.history.push('/site-admin/...')` instead: [example](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/web/src/enterprise/extensions/registry/RegistryNewExtensionPage.tsx#L103-104).

Also, declare them as methods on `OnboardingChecklist` rather than variables declared in `render()`, so that they're not re-declared every time the component renders.

beyang replied:
Thanks for the early feedback!

beyang replied:
I've found it actually pretty annoying to use `history`. If I have some child component that wants to reference `history.H` and the parent doesn't have `history` as one of its prop fields, I now need to walk up the component hierarchy adding `history` to each parent component prop. This feels funny, as I view "history" as a global thing like "window" that ought to be referencible anywhere. Is this just an idiosyncrasy of React / is there something I'm not getting?

sqs replied:
This is an explicit goal of react. It encourages you to make pure components, which are easier to test/refactor/etc.

beyang commented:
This is ready for initial design and code review.

## Design review request

@francisschmaltz @sqs there are two components to review:
* The dropdown: ![image](https://user-images.githubusercontent.com/1646931/52692352-86df1c00-2f18-11e9-9170-2b75074959f2.png)
* The site-admin panel: ![image](https://user-images.githubusercontent.com/1646931/52692361-8a72a300-2f18-11e9-843e-fca672f41c38.png)

Please check out the branch to local dev and review the following behaviors:
* Appearance of both components after initial setup (when no items are completed)
* Does clicking a checklist item do what you expect? Do you have enough context to complete the setup action?
* Behavior of the `(i)` info expansions—are these helpful?
* Any changes to the copy of either component
* Appearance of both components after complete setup (after all items are completed)
* Is the nav bar component prominent / noticeable enough to bug the site admin to complete the onboarding steps?

Note that the implementation diverges a bit from the design spec. I used icons from our icon set, added explanatory copy (similar to what the OpsGenie onboarding checklist has), moved the checklist icons from right to left, added the `(i)` buttons, and changed the text padding a bit.

## Code review request

@felixfbecker @lguychard, can I get a preliminary code review? A couple of questions in particular:
* Have I placed new components in the right directories?
* Are changes to existing components kosher?
* Does the .scss look okay?
* What amount of unit tests should I add?

sqs commented:
Looking awesome!

> Does clicking a checklist item do what you expect? Do you have enough context to complete the setup action?

I initially expected clicking a checklist item (**Enable an extension** specifically) to dismiss the popover. I understand it is helpful to have the popover still visible to remind me what I was supposed to do after I clicked.

> Behavior of the `(i)` info expansions—are these helpful?

Yes, and I like that they are short inline explanations. It would be nice if they linked to the most relevant doc page with full details.

> Any changes to the copy of either component

No significant copy changes that would change the length. I will push a commit to this branch (in the next ~90 min) with some minor copy changes.

> Appearance of both components after complete setup (after all items are completed)

It tells me that **Enable repositories** is not complete, despite my having enabled repositories:

![image](https://user-images.githubusercontent.com/1976/52746692-f562bf00-2f96-11e9-9cd0-1354ba17ff15.png)


> Is the nav bar component prominent / noticeable enough to bug the site admin to complete the onboarding steps?

I think so.

---

A couple things not accounted for yet:

- [ ] Needs a "Ignore remaining steps" ("Mark all as read") that makes this go away. It is OK to put this button only on the site admin overview page's component (so it's not *too* prominent).
- [ ] "Enable an extension" is not actually the step we want, because site admins don't need to enable any extensions to get activated (since code intel extensions are enabled by default). They need to perform a search with results or find-references. (The completion meter should list both separately.)
- [ ] The search-with-results and find-references steps should also be visible to non-site admin users. They would have a much shorter list of steps to take, but it would still be helpful for them. Also, for these users (and site admins), the completion meter should show up on their account profile page (like how it does on the site admin overview page).
- [ ] In the dark theme, hover background color of popover items makes them unreadable, and the top and bottom ones have rounded borders that seem unintentional. This should not happen if you are using standard components/colors/Bootstrap classes.

beyang commented:
@sqs thanks for the comments. Will address shortly, but 3 prelim questions:

> Needs a "Ignore remaining steps" ("Mark all as read") that makes this go away. It is OK to put this button only on the site admin overview page's component (so it's not _too_ prominent).

OK if this is added in a follow-up PR? This requires adding some backend state, which would make this a larger diff.

> "Enable an extension" is not actually the step we want, because site admins don't need to enable any extensions to get activated (since code intel extensions are enabled by default). They need to perform a search with results or find-references. (The completion meter should list both separately.)

Do we want to make enabling a non-code-intel extension part of activation?

> The search-with-results and find-references steps should also be visible to non-site admin users. They would have a much shorter list of steps to take, but it would still be helpful for them. Also, for these users (and site admins), the completion meter should show up on their account profile page (like how it does on the site admin overview page).

I think we got our wires crossed here a bit. We spoke about adding these in person, but the actual issue and design don't list these as part of the admin checklist. My impression from chatting with @francisschmaltz and @dadlerj was that these would be 2 separate checklists and that the site admin one only includes those covered in the design spec. What is the source of truth here?

sqs commented:
> "Ignore remaining steps" OK if this is added in a follow-up PR? This requires adding some backend state, which would make this a larger diff.

Yep.

> Do we want to make enabling a non-code-intel extension part of activation?

No.

> My impression from chatting with @francisschmaltz and @dadlerj was that these would be 2 separate checklists and that the site admin one only includes those covered in the design spec. What is the source of truth here?

The source of truth is https://docs.google.com/document/d/1P7uMHtOeYK1IwMj-1WtKspo16v8C_Mevj3qe1cX_kGY/edit#heading=h.pjjfcq8pvrce. Under **All users** it shows 3 things, and it says that it should be shown on the user profile page. I just got back from being out, so ideally I would have been around to provide this feedback earlier.

sqs commented:
Re: me updating the copy, I will wait for this to work for normal users too to do a holistic copy review. You can just use the copy from the gdoc for each step and use lorem ipsum or whatever you want to write as the detail text for each step.

beyang commented:
Design notes:
* Remove `(i)` buttons and replace in hover tooltips. In site admin panel and dropdown.
* Visual indication TBD. Should happen on first user action.
* Mental model: setup status (reflects the current setup completeness of the instance; it's possible to regress some steps); confetti can fall multiple times
* Gray and green, not red and green
* "Check all as done" button
* Toggle for removing from nav bar, but *always* show on site-admin and user profile
  * If a user has hidden it, but on shipping a new version we add another item, it should show up again

Q: ignore SSO step—toggle or button?
Q: three locations? site-admin, user profile, dropdown?
Q: hide nav bar when on site-admin page?

Beyang impl notes
* Need to add a way to trigger a status update (because need to show confetti when progress meter updates)
* Dismissable alerts key (dismiss the key, but the key changes for new Sourcegraph versions). Key on some hardcoded constant (that gets updated when new items are added).

lguychard commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+/**
+ * percentageDone returns the percent of activation checklist items completed.
+ */
+export const percentageDone = (info: SiteAdminChecklistInfo): number => {
```
```suggestion
export const percentageDone = (info?: SiteAdminChecklistInfo): number => {
    if (!info) {
    }
```

lguychard commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+ * SiteAdminActivationPopoverButton is the nav bar button that opens the site admin
+ * activation checklist as a dropdown.
+ */
+export class SiteAdminActivationPopoverButton extends React.PureComponent<
```
Regarding testing:
- You might find storybooks useful when iterating on this (you can check out the browser extension's storybooks: `cd client/browser && yarn run storybook`)
- You should still add snapshot tests for the various states of this component
- To this end, `fetchSiteAdminChecklist` should be injected into this component's props, not referenced directly, so that actually making a graphQL request is not needed to test your component (eg. in tests, you can do `typescript <SiteAdminActivationPopoverButton fetchSiteAdminChecklist={of(/* some dummy checklist data */)} .../>`
- Here's an example of a recent PR that adds snapshot tests to a component: https://github.com/sourcegraph/sourcegraph/pull/2280

lguychard commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+}
+
+/**
+ * ChecklistItem is a single item in the site admin activation checklist.
```
Consistency (very minor): in most of the TS codebase, the docstrings do not repeat the function/class/method name.

lguychard commented on web/src/site-admin/SiteAdminActivation.scss:
```diff
@@ -0,0 +1,94 @@
+}
+
+.onboarding-button {
+    .link-like {
```
You're probably getting lint warnings about this: generally, we use BEM modifiers and not descendant selectors: [example](https://sourcegraph.com/search?q=r:sourcegraph/sourcegraph+f:.scss+/%26__/).

felixfbecker commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+
+    private toggleDetail = (e: React.MouseEvent<HTMLSpanElement>) => {
+        e.stopPropagation()
+        this.setState({ showDetail: !this.state.showDetail })
```
Never reference `this.state` in `setState()` input because it is async. Use the callback form
```suggestion
        this.setState(state => ({ showDetail: !state.showDetail }))
```

felixfbecker commented on web/src/site-admin/SiteAdminOverviewPage.tsx:
```diff
@@ -93,13 +103,17 @@ export class SiteAdminOverviewPage extends React.Component<Props, State> {
             fetchWeeklyActiveUsers().subscribe(stats => this.setState({ stats }), error => this.setState({ error }))
         )
+        this.subscriptions.add(
+            fetchSiteAdminChecklist().subscribe(onboardingChecklist => this.setState({ onboardingChecklist }))
```
This swallows the error if the request fails

felixfbecker commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+        this.subscriptions.add(
+            fetchSiteAdminChecklist().subscribe(checklistInfo => {
+                this.setState({ checklistInfo })
+            })
```
This swallows the error if the request fails. The user should be informed that something went wrong.

felixfbecker commented on web/src/nav/NavLinks.tsx:
```diff
@@ -64,6 +66,15 @@ export class NavLinks extends React.PureComponent<Props> {
                 />
+                {this.props.authenticatedUser && this.props.authenticatedUser.siteAdmin && (
+                    <SiteAdminActivationPopoverButton
+                        menu={ContributableMenu.CommandPalette}
```
Is this correct?

ijsnow commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+ * SiteAdminActivationPopoverDropdown presents the site admin activation checklist
+ * in a navbar dropdown.
+ */
+export class SiteAdminActivationPopoverDropdown extends React.PureComponent<SiteAdminChecklistProps, {}> {
```
It should also be in it's own file. Same for each component in this file. If it truly is worth splitting into it's own component, it should be in it's own file.

ijsnow commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+
+type SiteAdminActivationPopoverButtonProps = CommandListProps & { history: H.History }
+interface SiteAdminActivationPopoverButtonState {
+    checklistInfo?: SiteAdminChecklistInfo
```
The fact that the state only has one property and it is itself an object is strange to me. I'd make each field of `checkListInfo` a property of the state itself.

ijsnow commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+
+interface SiteAdminChecklistProps {
+    history: H.History
+    checklistInfo?: SiteAdminChecklistInfo
```
As with the state above, I'd just have `SiteAdminChecklistProps` extend `SiteAdminChecklistInfo`.

ijsnow commented on web/src/site-admin/SiteAdminActivation.tsx:
```diff
@@ -0,0 +1,259 @@
+                        <InformationOutlineIcon className="icon-inline" />
+                    </span>
+                </p>
+                <p className={'item-detail' + (this.state.showDetail ? ' show' : ' hide')}>{this.props.detail}</p>
```
Not necessary, but you could also use a template string
```suggestion
                <p className={`item-detail ${this.state.showDetail ? 'show' : 'hide'}`}>{this.props.detail}</p>
```

dadlerj commented on cmd/frontend/internal/pkg/usagestats/usage_stats.go:
```diff
@@ -474,6 +488,8 @@ func LogActivity(isAuthenticated bool, userID int32, userCookieID string, event
 		return logSearchQuery(userID)
 	case "PAGEVIEW":
 		return logPageView(userID)
+	case "CODEINTEL:REFS":
```
Beat you to a change here https://github.com/sourcegraph/sourcegraph/pull/2292/files... Sorry, wasn't aware you were adding this, but it appears our changes were identical. Cogit (or whatever) would have been useful here!

beyang commented:
@sqs PTAL. 3 things to check:
* The "Setup" button in the navbar
* The site-admin panel
* The user account profile panel

When user is site admin, list has 5 items; when not, list has 2 items.

@felixfbecker @lguychard PTAL at the code. I've split this up into smaller commits and added tests for each component. The main "model" class is [`ActivationStatus`](https://github.com/sourcegraph/sourcegraph/pull/2258/files#diff-41c5e976719d99f04f96fe93fab15eaaR15), which is the source of truth on the frontend of which activation steps have been completed. The new UI components are:
* `ActivationChecklist`, which renders a checklist of all activation items
* `ActivationPopoverButton`, which renders the nav bar item, which displays a circular progress bar of activation progress
* `ActivationClickTarget`, which wraps another component and triggers activation status updates when it is clicked.

lguychard commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+     * Accepts the activation steps to track and a function that fetches the completed
+     * activation steps (in the form of a function that returns an Observable).
+     */
+    constructor(steps: ActivationStep[], fetchCompleted: () => Observable<{ [key: string]: boolean }>) {
```
```suggestion
    constructor(private steps: ActivationStep[], private fetchCompleted: () => Observable<{ [key: string]: boolean }>) {
```

This allows you to remove the declarations `steps_` and `fetchCompleted_` above.

lguychard commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+export class ActivationStatus {
+    private steps_: ActivationStep[]
+    private fetchCompleted: () => Observable<{ [key: string]: boolean }>
+    private completed_: BehaviorSubject<{ [key: string]: boolean } | null>
```
```suggestion
    private completed = new BehaviorSubject<{ [key: string]: boolean } | null>(null)
```

lguychard commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    public get steps(): ActivationStep[] {
+        return this.steps_
```
No need for this, just declare `public readonly` properties.

lguychard commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+            if (!this.completed.value) {
+                return
+            }
+            const newVal: { [key: string]: boolean } = {}
```
```suggestion
            const newVal = { ...this.completed.value }
```

lguychard commented on shared/src/components/activation/ActivationClickTarget.tsx:
```diff
@@ -0,0 +1,101 @@
+                        return
+                    }
+                    // Completed list is available, so now determine whether it's activated by looking at what's completed
+                    let activated = true
```
```suggestion
                    const activated = this.props.activationKeys.every(k => completed[k] !== undefined)
```

removes the need for the `for` loop below

lguychard commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+     * Must be called with null parameter first (this triggers the first fetch of
+     * the completion status via `fetchCompleted`).
+     */
+    public update(u: { [key: string]: boolean } | null): void {
```
```suggestion
    public update(activatePatch: { [key: string]: boolean } | null): void {
```

For consistency with the code in `ActivationClickTarget` + to avoid having a single-letter variable name

lguychard commented on web/src/tracking/activation.tsx:
```diff
@@ -0,0 +1,153 @@
+            }
+        }
+    `).pipe(
+        map(dataAndErrors),
```
Why is this needed instead of just `map(({ data, errors }) => {` below?

sqs commented:
Running through it now, will add notes:

- [ ] Font color of `Setup` differs from the font color of the other items in the global navbar in both light and dark themes. Should be the same. 
![image](https://user-images.githubusercontent.com/1976/53142707-c970ac00-3549-11e9-831e-f36a767fbd98.png)

sqs commented on web/src/tracking/activation.tsx:
```diff
@@ -0,0 +1,153 @@
+            {
+                id: 'didSearch',
+                title: 'Search your code',
+                detail: 'Issue a search query over your code.',
```
```suggestion
                detail: 'Perform a search query on your code.',
```

Recommend not using "issue" as a verb; see **Avoid ambiguous verbs** at https://sourcegraph.com/github.com/sourcegraph/about@master/-/blob/STYLEGUIDE.md#clarity.

felixfbecker commented on shared/src/components/activation/Activation.scss:
```diff
@@ -0,0 +1,127 @@
+    animation: bottomBubbles ease-in-out 0.75s forwards;
+}
+
+@keyframes topBubbles {
```
Note that this is a global name, you should probably give it a name that's globally unambiguous

felixfbecker commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+        return this.completed_
+    }
+
+    public ensureInit(): void {
```
Does this method always have to be called to bring the object into a valid state? If yes, why not call it in the constructor?

felixfbecker commented on shared/src/components/activation/Activation.scss:
```diff
@@ -0,0 +1,127 @@
+        padding: 0;
+        margin: 0;
+    }
+    li {
```
Don't use child selectors (see lint warning).
```suggestion
    &__checklist-item {
```

felixfbecker commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+ */
+export class ActivationStatus {
+    private steps_: ActivationStep[]
+    private fetchCompleted: () => Observable<{ [key: string]: boolean }>
```
I am confused by the use of string keys/IDs everywhere in this code. Aren't the completion steps well-known? If so, this should be made type safe

beyang replied:
To do that, I'd do the following:
* Define an empty interface `ActivationSteps` under `shared/`
* Extend the interface under `web/`
* Extend `ActivationStatus` under `web/` where the extended class uses the extended `ActivationSteps` interface

This seemed more verbose than justified for now. WDYT?

felixfbecker replied:
Why the split between `web` and `shared`? And why an empty interface? Can there be different sets of steps?

I seem to misunderstand something here. My assumption is these activation steps are only shown in the webapp and coupled to an instance (server-side data).

beyang replied:
The split is necessary, because the `ActivationProps` need to be in shared, because there are certain UI components like `Panel` that can be activation triggers. We also might want to add activation to the browser extension in the future.

felixfbecker replied:
That explains why one part has to be in shared/, but why does the non-empty interface have to be in web? 

Can there be different sets of steps?

beyang replied:
Yeah, the activation steps are specific to the application. For instance, "configure SSO" and "add repositories" make less sense for browser ext.

felixfbecker commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Tracks the activation status of the current user.
+ */
+export class ActivationStatus {
```
It's unusual to use classes to pass state between React components. Usually you just pass the data as props or a prop with a plain object, which means React can automatically rerender components when the data changes without having to listen to any Observables manually. To update data, you pass around an `onFooChange` function alongside the data. The highest common ancestor then manages this state.

beyang replied:
How would that look with the activation state? Something like:
* Define `ActivationProps` under `shared/`. This would include all the fields currently under `ActivationStatus` including callbacks.
* Have a top-level component just under the `SourcegraphWebApp` instance and manages the activation state.
* Pass through all `ActivationProps` down the component hierarchy to get to the UI components that may trigger activation.

This seems less well-encapsulated than the current code (because then activation props would be spread across many components).

felixfbecker commented on shared/src/components/activation/ActivationClickTarget.tsx:
```diff
@@ -0,0 +1,101 @@
+        this.state = {}
+    }
+    public componentDidMount(): void {
+        if (this.props.activation) {
```
You are not handling props changes of `this.props.activation` itself (i.e. through `componentDidUpdate()`. (note this would be eliminated if you just passed the data as props directly)

felixfbecker commented on shared/src/components/activation/ActivationClickTarget.tsx:
```diff
@@ -0,0 +1,101 @@
+        }
+        if (this.state.activated) {
+            return
+        }
```
```ts
if (!this.props.activation || this.state.activated !== false) {
    return
}
```

felixfbecker commented on shared/src/components/activation/ActivationClickTarget.tsx:
```diff
@@ -0,0 +1,101 @@
+            e.stopPropagation()
+            setTimeout(this.props.pauseAndRetrigger, 1000)
+        }
+        this.setState({ activated: true, animate: !this.state.activated })
```
Never reference `this.state` in `setState()` input, `setState()` is async.

```suggestion
        this.setState(state => ({ activated: true, animate: !state.activated }))
```

felixfbecker commented on shared/src/components/activation/ActivationPopoverButton.tsx:
```diff
@@ -0,0 +1,111 @@
+                        }`}
+                    >
+                        <CircularProgressbar
+                            styles={{
```
You should use CSS, not inline styles, like the rest of our codebase. CircularProgressBar supports a `className` prop, and you can set `stroke` with CSS. See https://github.com/iqnivek/react-circular-progressbar#using-css

felixfbecker commented on shared/src/graphql/graphql.ts:
```diff
@@ -40,6 +40,15 @@ export function dataOrThrowErrors<T extends GQL.IQuery | GQL.IMutation>(result:
     return result.data
 }
 
+export function dataAndErrors<T extends GQL.IQuery | GQL.IMutation>(
```
This function just copies its input so I don't see a point in it?

felixfbecker commented on web/src/search/input/SearchButton.tsx:
```diff
@@ -3,34 +3,54 @@ import HelpCircleOutlineIcon from 'mdi-react/HelpCircleOutlineIcon'
     /** Never show the "Search" button label. */
     noLabel?: boolean
+
+    activation?: ActivationStatus
```
Tip, if you want to avoid copy-pasting this props everywhere, define an interface `ActivationStatusProps` and make the Props interfaces extend it. Also puts documentation for the prop in a single place and makes it easy to modify the type.

felixfbecker commented on web/src/search/input/SearchButton.tsx:
```diff
@@ -3,34 +3,54 @@ import HelpCircleOutlineIcon from 'mdi-react/HelpCircleOutlineIcon'
-    public state: State = { isOpen: false }
+    public state: State = { isOpen: false, didClick: false }
+
+    private button: React.RefObject<HTMLButtonElement>
```
```suggestion
    private button = React.createRef<HTMLButtonElement>()
```

Prefer initialisers over constructors.

felixfbecker commented on web/src/search/input/SearchButton.tsx:
```diff
@@ -3,34 +3,54 @@ import HelpCircleOutlineIcon from 'mdi-react/HelpCircleOutlineIcon'
+        this.button = React.createRef()
+    }
+
+    private retrigger = () => this.button.current && this.button.current.click()
```
It seems unneeded to me to programmatically click the button with React. Refs are to be avoided in React: https://reactjs.org/docs/refs-and-the-dom.html#dont-overuse-refs

I guess this is to trigger a submit event listener that is defined higher up, so you could pass a `submit` callback prop down here that you can call to also trigger this event listener.

felixfbecker commented on web/src/search/input/SearchButton.tsx:
```diff
@@ -3,34 +3,54 @@ import HelpCircleOutlineIcon from 'mdi-react/HelpCircleOutlineIcon'
-                </button>
+                <ActivationClickTarget
+                    activation={this.props.activation}
+                    activationKeys={['didSearch']}
```
Don't pass inline array

felixfbecker commented on web/src/site-admin/SiteAdminOverviewPage.tsx:
```diff
@@ -93,13 +100,23 @@ export class SiteAdminOverviewPage extends React.Component<Props, State> {
+        if (this.props.activation) {
+            this.subscriptions.add(
+                this.props.activation.completed.subscribe(completed => {
+                    if (completed) {
```
Shouldn't it also be set to `false` if `completed === false`?

felixfbecker commented on web/src/site-admin/SiteAdminOverviewPage.tsx:
```diff
@@ -117,6 +134,35 @@ export class SiteAdminOverviewPage extends React.Component<Props, State> {
+                                    title={`${setupPercentage}% of setup completed`}
+                                    defaultExpanded={setupPercentage < 100}
+                                >
+                                    <div className="activation-container">
```
You shall not use a class name defined for a different component in another component. This should be `site-admin-overview-page__activation-container`, defined in SiteAdminOverviewPage.scss

felixfbecker commented on web/src/tracking/activation.tsx:
```diff
@@ -0,0 +1,153 @@
+    `).pipe(
+        map(dataAndErrors),
+        map(dataAndErrors => {
+            if (!dataAndErrors.data) {
```
`data` is `null` if there was an error, in which case you shouldn't discard what is in `errors` but somehow propagate it and show it in the UI (in the common/simple case with `dataOrThrowErrors` and catching the error in the UI)

felixfbecker commented on web/src/tracking/activation.tsx:
```diff
@@ -0,0 +1,153 @@
+            }
+            const rURLs = data.repositories.nodes
+                .filter(r => r.gitRefs && r.gitRefs.totalCount > 0)
+                .sort(r => (r.gitRefs ? -r.gitRefs.totalCount : 0))
```
Sort receives two elements of the array `(a, b)` that you need to compare. If you want to sort by projected values, you can use `sortBy()` from lodash.

felixfbecker commented on web/src/tracking/activation.tsx:
```diff
@@ -0,0 +1,153 @@
+            if (!data.repositories.nodes) {
+                return null
+            }
+            const rURLs = data.repositories.nodes
```
```suggestion
            const repositoryUrls = data.repositories.nodes
```

We like readable variable names in TypeScript :)

felixfbecker commented on web/src/tracking/activation.tsx:
```diff
@@ -0,0 +1,153 @@
+ *
+ * @param isSiteAdmin determines if site-admin-only activation steps are included.
+ */
+export const newActivationStatus = (isSiteAdmin: boolean) => {
```
Factory functions in TS/JS are named `create*` because functions should start with an imperative verb.
```suggestion
export const createActivationStatus = (isSiteAdmin: boolean) => {
```

beyang commented:
@sqs @francisschmaltz this is ready for another round of design review. I've addressed all of @sqs's comments except for the button animation. The current state of the branch displays the previous animation, which looks cool but will be unnoticed if not more prominent. @francisschmaltz if you'd like to test out a new animation in the context of the UI, check out this branch and change the `.first-use-button` selectors in `Activation.scss`.

Asks:
* @francisschmaltz Can you send me updated CSS for the animation (or we can just use https://daniel-lundin.github.io/react-dom-confetti/ or another off-the-shelf animation)
* @sqs Can you take another pass and list out any issues / changes you'd like to see? (Note: you must currently be site admin user to test it out.) If everything else on UX LGTY, then gimme a LGTM.

Note: after I get sign-off on the design, I will go back and re-clean-up the code. Change requests after I add tests, etc. will be about 10x more expensive, so please get in all the design/behavior feedback now.

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -147,6 +170,62 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+            return
+        }
+        const fetcher = fetchActivationStatus(authenticatedUser.siteAdmin)
+        const firstFetched = new Promise<void>(resolve => {
```
You almost never need the Promise constructor. Use `.toPromise()` to convert an Observable to a Promise

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -147,6 +170,62 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+            .subscribe(c => this.setState({ activationCompleted: c }))
+    }
+
+    private updateActivation = (u: { [key: string]: boolean }) => {
```
```suggestion
    private updateActivation = (update: { [key: string]: boolean }) => {
```

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -147,6 +170,62 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+        return {
+            steps: getActivationSteps(authenticatedUser.siteAdmin),
+            fetcher,
+            firstFetched,
```
It's unusual to put something like fetcher or a Promise into state. State holds the data that the component renders (and the component will rerender whenever it changes). Those two things don't really belong in that category.

I think you should not need a "fetcher", since you can always just pass `siteAdmin` at call time from `this.state.authenticatedUser`.

The Promise you can save as a private property.

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -236,3 +325,144 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+            siteAdminOnly: true,
+        },
+    ]
+        .filter(e => true || !e.siteAdminOnly)
```
This is always `true`

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -147,6 +170,62 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+            return
+        }
+        this.state.activation
+            .fetcher()
```
This would trigger multiple HTTP requests even if there currently is one in flight, and if that one is slower than the new request, then the state can get overridden with outdated data.
I would recommend you use Rx for this.
Consider the "requests to refresh" a stream, represented as a Subject in a private property.
Then you just `switchMap()` those in `componentDidMount()` to fetching the activation status (reading `this.state.authorizedUser` directly) and set the state in the subscribe.
You trigger this Subject with `.next()` in 1) `componentDidMount()` and 2) the function you pass to child components to request an update.

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -236,3 +325,144 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+    steps: ActivationStep[]
+}
+
+const fetchActivationStatus = (isSiteAdmin: boolean) => () =>
```
Move these functions into a separate file to not clutter this component

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -216,6 +295,16 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+                                            this.state.activation &&
+                                            this.state.activationCompleted && {
+                                                steps: this.state.activation.steps,
+                                                update: this.updateActivation,
```
I think there is a problem here in only passing the `update` functions if `this.state.activation` is defined. `this.state.activation` will be `undefined` while it's still loading. But during that time, the user can already trigger actions that would update the status. But they can't report them, because they don't receive the `update` function yet, and the event will get dropped. See this condition: https://github.com/sourcegraph/sourcegraph/pull/2258/files#diff-e34d503c07aa5c46198ea10993381e4eR29

Instead, they should always get the `update` function, so that these updates are "queued up" (through attaching to the Promise).

beyang replied:
Nice catch, I'll make the activation defined even if the completion status has not yet loaded.

felixfbecker commented on yarn.lock:
```diff
@@ -12537,6 +12572,16 @@ react@^16.6.3, react@^16.7.0:
+    loose-envify "^1.1.0"
+    object-assign "^4.1.1"
+    prop-types "^15.6.2"
+    scheduler "^0.13.2"
```
This introduced a lot of duplicate dependencies. Try running `npx yarn-deduplicate`

felixfbecker commented on web/src/search/input/SearchButton.tsx:
```diff
@@ -14,20 +15,27 @@ interface Props {
 interface State {
     isOpen: boolean
+
+    didClick: boolean
```
Unused?

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -236,3 +325,144 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+        },
+    ]
+        .filter(e => true || !e.siteAdminOnly)
+        .map(e => pick<any, keyof ActivationStep>(e, 'id', 'title', 'detail', 'action'))
```
```suggestion
        .map(({ siteAdminOnly, ...step }) => step)
```

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -236,3 +325,144 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+            id: 'didSearch',
+            title: 'Search your code',
+            detail: 'Perform a search query on your code.',
+            action: (h: H.History) => h.push('/search'),
```
It looks like the majority of these items are links. Whenever something is a link, it should be an actual `<a>` instead of using `onClick` and `history.push()`, so that opening in a new tab etc works. Could this be turned into real links, e.g. by adding a `to` option?

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -236,3 +325,144 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+        })
+    )
+
+const fetchReferencesLink: () => Observable<string | null> = () =>
```
```suggestion
const fetchReferencesLink = (): Observable<string | null> =>
```

felixfbecker commented on web/src/SourcegraphWebApp.tsx:
```diff
@@ -236,3 +325,144 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
+        })
+    )
+
+const getActivationSteps = (isSiteAdmin: boolean) =>
```
```suggestion
const getActivationSteps = (isSiteAdmin: boolean): ActivationStep[] =>
```

felixfbecker commented on shared/src/components/activation/ActivationDropdown.tsx:
```diff
@@ -0,0 +1,136 @@
+        if (newDone > oldDone && !this.state.animate) {
+            this.setState({ animate: true })
+            this.removeAnimationTimeout = setTimeout(() => {
+                this.setState({ animate: false })
```
This will result in weird behaviour when a prop update comes in that causes an animation to be triggered, then another one comes in before the timeout was finished.

Consider this code which handles such races (by queuing animations):
```ts
componentDidMount() {
	this.subscriptions.add(
	  this.componentUpdates.pipe(
	    map(props => props.activationStatus),
	    distinctUntilChanged((a, b) => isEqual(a, b)),
	    concatMap(() => concat([true], of(false).pipe(delay(1000))))
	  ).subscribe(addClass => {
	    this.setState({ addClass })
	  })
	)
}
componentDidUpdate() {
    this.componentUpdates.next(this.props)
}
```

beyang commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * The completion status of each activation step
+     */
+    completed?: { [key: string]: boolean }
```
Don't think that's the solution here, because you still want to have a callback `update` that updates status by ID. I think the solution you propose here is better: https://github.com/sourcegraph/sourcegraph/pull/2258/files#r261230964

beyang replied:
After attempting to genericize this code (see the `bl/setup-generic` branch), I've concluded that making this generic would add a lot of bloat/verbosity to the code that doesn't justify the benefit of type safety here. Here's the reasoning:
* It is convenient to have completion status stored as a map of ID to boolean, instead of within the `ActivationStep` interface. That makes intuitive sense and means we can rely on that value being defined/undefined as an indicator of whether it has been fetched from the server. If I moved it into `ActivationStep`, that would mean letting the constraints of the type system dictate the layout of the data structure—I don't think this is a good tradeoff.
* Given that the completion status is stored as a separate map, there isn't a way to genericize this without adding a lot of verbosity. Basically, it would mean having to parameterize every type under `shared/` that holds a reference to `Activation` or `ActivationStep` (see the `bl/setup-generic` branch for an attempt).

felixfbecker replied:
Did you forget to push that branch? I cannot find it

beyang replied:
Pushed to `bl/setup-generic`. Another thought here we can just use string constants for a higher degree of safety w/o the overhead of using the type system.

felixfbecker replied:
I would at least like the keys as constants / string enum defined somewhere. It is currently non-trivial to understand what keys even exist and to make sure you're not mistyping one

beyang commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,54 @@
+    id: string
+    title: string
+    detail: string
+    action: (h: H.History) => void
```
It is necessary to pass the history as a param, because the step is defined outside a UI context that has a history (at the top-level component `SourcegraphWebApp`), but must be executed within a context that has a history item. I don't see how the action could close over a `History` instance without moving the definition of the activation steps to a lower component (`Layout` or lower) and conflating it with some specific UI.

beyang commented:
@felixfbecker PTAL

felixfbecker commented on shared/src/components/activation/ActivationDropdown.tsx:
```diff
@@ -32,16 +34,24 @@ export class ActivationDropdown extends React.PureComponent<Props, State> {
+            .pipe(
+                map(props => props.activation.completed),
+                distinctUntilChanged((a, b) => {
+                    console.log(a, b)
```
remove
```suggestion
```

beyang commented:
@felixfbecker PTAL. I moved the activation state management into a separate class `ActivationStatus`, which handles synthesizing updates and refetches into a single Observable that emits the completed status. The `SourcegraphWebApp` component then subscribes to this and updates its own state accordingly.

beyang commented on web/src/tracking/ActivationStatus.tsx:
```diff
@@ -0,0 +1,227 @@
+     * statuses.
+     */
+    public get completed(): Observable<ActivationCompleted | null> {
+        return this.completed_
```
That behavior is quite surprising to me. Does that mean the following code also suffers from the same bug? https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/management-console/web/src/MonacoEditor.tsx#L86-87

In the link, `componentUpdates` is a Subject that is unsubscribed in `componentWillUnmount`. If the component is then remounted, wouldn't the Subject then be invalid?

beyang commented:
@felixfbecker I removed `ActivationStatus` and replaced it with a HOC constructor, `withActivation`. PTAL.

Also, does my question [here](https://github.com/sourcegraph/sourcegraph/pull/2258#discussion_r262285402) make sense?

felixfbecker commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,277 @@
+                    )
+                    .subscribe(completed => {
+                        this.setState({ completed })
+                        this.completionStatusSet.next()
```
Triggering a Subject from the subscribe of another Observable is usually a red flag that the Subject should be replaced by unifying everything into one Observable chain. That is easier to reason about, because there is a uni-directional data/control flow, whereas cross-triggering Subjects quickly becomes an entangled mess.

Here's a proposal how to do everything in a single Observable chain, where the Subjects are only used to kick the chain off from the callbacks.

```ts
this.subscriptions.add(
    combineLatest(
        // Completion status on the server
        combineLatest(
            // Refetch completion status from server when user changes
            this.componentUpdates.pipe(
                map(props => props.authenticatedUser),
                distinctUntilChanged()
            ),
            // Refetch completion status from server when requested
            this.refetches
        ).pipe(
            switchMap(([authenticatedUser]) =>
                authenticatedUser ? fetchActivationStatus(authenticatedUser.siteAdmin) : []
            )
        ),
        // Local completion status updates
        this.completionStatusUpdates.pipe(
            tap(update => {
                // Send update to server for events that don't themselves trigger an update
                if (update.FoundReferences) {
                    logUserEvent(GQL.UserEvent.CODEINTELREFS)
                }
            }),
            scan((prevLocalStatus, update) => ({ ...prevLocalStatus, ...update }))
        )
    )
        .pipe(
            // Always rebase local completion status updates on top of server completion status
            map(([serverCompletionStatus, localCompletionStatus]) => ({
                ...serverCompletionStatus,
                ...localCompletionStatus,
            }))
        )
        .subscribe(completed => {
            this.setState({ completed })
        })
)
```

felixfbecker commented on shared/src/components/activation/Activation.tsx:
```diff
@@ -0,0 +1,84 @@
+/**
+ * A map indicating which activation steps have been completed
+ */
+export type ActivationCompletionStatus = { [K in ActivationID]?: boolean }
```
I see that every key is optional (possibly `undefined`), meaning there are three completion states for each step. What is the semantic of the `undefined` state? Since they are always all loaded together, shouldn't they always be either `true` or `false`?

beyang replied:
The activation steps change depending on whether a user is site admin or not. (The site admin currently has 5 steps to complete vs. 2 for non-admin users.) So we can't expect every completion object to have all 5 fields. Perhaps there is a way to encode this in the type system, but it feels like more work than warranted at this point, esp given the activation logic may change a fair bit.

beyang commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,277 @@
+
+        private updateCompletionStatus = (update: ActivationCompletionStatus): void => {
+            this.completionStatusSet.pipe(first()).subscribe(() => {
+                this.recordUpdate(update)
```
The `recordUpdate` call can be moved outside the `completionStatusSet` subscription block, but I don't think it makes that much of a difference. It will only affect when the update is persisted to the server (which is either before or after the initial completion fetch). This has no direct effect on the activation status displayed to the user.

beyang commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,277 @@
+}
+
+export const withActivation = <P extends ActivationProps & Props>(Component: React.ComponentType<P>) =>
+    class WithActivation extends React.Component<Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>, State> {
```
There's also the `Subtract<T, K>` type from [`utility-types`](https://github.com/piotrwitek/utility-types). I noticed that neither that nor `type-zoo` is a dependency yet. Is it worth adding / which package would you prefer?

felixfbecker replied:
utility-types looks more complete 👍🏻

beyang replied:
Will do this in a follow-up PR (https://github.com/sourcegraph/sourcegraph/issues/2634).

beyang commented:
@felixfbecker PTAL

felixfbecker commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,271 @@
+                map(props => props.authenticatedUser),
+                distinctUntilChanged()
+            )
+            const serverCompletionStatus: Observable<ActivationCompletionStatus> = combineLatest(
```
Not blocking, I would usually recommend not saving Observables in a variable because it makes it possible to reference them in multiple places (which would make the data flow a fork), whereas if there are inlined it's clear the data flow is linear. Unless you _need_ to fork of course.

beyang replied:
I think there's some tension here, because IMO the unnamed/linear version of the code was hard for me to decipher. Maybe this will become easier the more familiar I get with Observables, but for now I'd prefer to leave it as is, with relatively simple expressions and named variables.

felixfbecker commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,271 @@
+            ).pipe(
+                tap(update => update && recordUpdate(update)),
+                scan((prev, next) => (next ? { ...prev, ...next } : {})),
+                map(p => p || {})
```
I think `p` can't be `null` at this point anymore, you just need a type declaration above to help the inference.
```suggestion
```

felixfbecker commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,271 @@
+                this.updates
+            ).pipe(
+                tap(update => update && recordUpdate(update)),
+                scan((prev, next) => (next ? { ...prev, ...next } : {})),
```
```suggestion
                scan<Partial<ActivationCompletionStatus> | null, Partial<ActivationCompletionStatus>>(
                    (prev, next): Partial<ActivationCompletionStatus> => (next ? { ...prev, ...next } : {})
                ),
```

felixfbecker commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,271 @@
+export const withActivation = <P extends ActivationProps & Props>(Component: React.ComponentType<P>) =>
+    class WithActivation extends React.Component<Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>, State> {
+        private subscriptions = new Subscription()
+        private componentUpdates = new Subject<Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>>()
```
```suggestion
        private componentUpdates = new Subject<Readonly<Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>>>()
```

felixfbecker commented on web/src/tracking/withActivation.tsx:
```diff
@@ -0,0 +1,271 @@
+        }
+
+        public componentDidUpdate(): void {
+            this.componentUpdates.next(this.props as Props & Pick<P, Exclude<keyof P, keyof ActivationProps>>)
```
I think you can get rid of this cast if you make the Subject readonly (see above)
```suggestion
            this.componentUpdates.next(this.props)
```
